src/code/z_map_data.c

```diff
@@ -0,0 +1,389 @@
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  22, 25,  29, 30, 32, 35,
+    39, 41, 23, 26, 23, 26, 31, 40, 42, 40, 24, 27, 24, 27, 28, 34, 36, 37, 38, 43, 33, 23, 26, 3,  8,   12, 14, 15, 31,
+    9,  22, 25, 30, 39, 21, 39, 0,  1,  4,  5,  6,  10, 17, 20, 0,  28, 30, 29, 31, 23, 5,  4,  15, 0,   0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,   0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  22, 23, 24, 25, 26, 9,  12, 14, 21, 8,  0,  0,  0,  0,  0,  0,   0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,   0,  0,  0,  0,
+    0,  0,  7,  8,  9,  8,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,   0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  255, 0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,   0,  0,  0,  0,
+    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
+};
+
+static u8 sFloorNames[10][8] = {
+    /* clang-format off */
+    {    0,    0,    0, F_3F, F_2F, F_1F, F_B1, F_B2 }, 
+    {    0,    0,    0,    0,    0,    0, F_2F, F_1F },
+    {    0,    0,    0,    0,    0,    0, F_1F, F_B1 },          
+    {    0,    0,    0,    0, F_2F, F_1F, F_B1, F_B2 },
+    {    0,    0,    0, F_5F, F_4F, F_3F, F_2F, F_1F }, 
+    {    0,    0,    0,    0, F_3F, F_2F, F_1F, F_B1 },
+    {    0,    0,    0,    0, F_4F, F_3F, F_2F, F_1F },    
+    {    0,    0,    0,    0, F_B1, F_B2, F_B3, F_B4 },
+    {    0,    0,    0,    0,    0, F_B1, F_B2, F_B3 },       
+    {    0,    0,    0,    0,    0,    0,    0, F_1F },
+}; /* clang-format on */
+
+/* Y coord of big skull icon on map screen, relative to center of screen.
+   -99 if no dungeon map, otherwise (51 - 14 * FloorNumber) */
+static s16 sSkullFloorIconY[10] = { -47, -47, -33, -47, -47, -5, -19, -47, -99, -99 };
+
+MapData gMapDataTable = {
+    sFloorTexIndexOffset,
+    sBossFloor,
+    sRoomPalette,
+    sMaxPaletteCount,
+    sPaletteRoom,
+    sRoomCompassOffsetX,
+    sRoomCompassOffsetY,
+    sDgnMinimapCount,
+    sDgnMinimapTexIndexOffset,
+    sOwMinimapTexSize,
+    sOwMinimapTexOffset,
+    sOwMinimapPosX,
+    sOwMinimapPosY,
+    sOwCompassInfo,
+    sDgnMinimapTexIndexBase,
+    sDgnCompassInfo,
+    sOwMinimapWidth,
+    sOwMinimapHeight,
+    sOwEntranceIconPosX,
+    sOwEntranceIconPosY,
+    sOwEntranceFlag,
+    sFloorCoordY,
+    sSwitchEntryCount,
+
+    /* These are declared as 2D arrays, but each requires a 2-byte null terminator,
+       so, we define them as 1D arrays here and cast the pointer. */
+    (void*)sSwitchFromRoom,
+    (void*)sSwitchFromFloor,
+    (void*)sSwitchToRoom,
```

Comment:
This bothers me but I guess I just don't really understand the issue. I don't think arrays of u8 arrays should introduce any alignment or padding, so what's this about 2-byte null terminators?

Reply:
It might be padding. Each of these three arrays has 51 bytes * 10 dungeons = 510 bytes, but then in the ROM it's followed by two zero bytes to make 512.

So, if we defined these arrays as u8[10]51], we'd need a separate u8[2] after it. At one point I did have it like that, but the linker eats it or something because it breaks matching, and I didn't investigate any further.

Reply:
It should always be at least 4 bytes aligned since it's in .data, regardless of types.

Reply:
Ah, you're right! The compiler does add that padding after all. Much cleaner now!